### DISCUSSION TITLE
Is `use` basically an `include`?

### DISCUSSION THREAD

**Discussion by theAkito** [nushell/nushell#7000] [Q&A] [answered]
The documentation regarding `use` says, that it "imports" a module.

https://www.nushell.sh/book/commands/use.html

However, when I "import" a module into `config.nu`, as seen in the following code block,

```nu
use /home/user/.config/nushell/master.nu *
```

then it is automatically exported into the environment sourcing the `config.nu`.

Shouldn't there be an `export use` statement?

**Comment by kubouch** [comment, marked as answer]
`use` imports symbols you select from a module to the current namespace (and runs the `export-env` block if there is one). It does not source the whole file.

`export use` exists already.

**Reply by theAkito** [reply]
Thanks, I would like a clarification on the following.

```nu
use /home/user/.config/nushell/master.nu *
```

When I have this line in `config.nu` and I source `config.nu` I get all the definitions from `/home/user/.config/nushell/master.nu`, as well. So, that's already "exported" isn't it?

**Reply by kubouch** [reply]
You get only those definitions that were specified with `export` in the `master.nu`. Also, `use` does not evaluate the module (except the `export-env` block). When you `source` a script, it evaluates the whole file and brings everything into the local scope.

**Reply by theAkito** [reply]
> You get only those definitions that were specified with export in the `master.nu`.

I thought, this export was meant for the module, which `use`s the module directly, rather than all subsequent environments sourcing the module, which is `use`ing `master.nu`.

**Reply by theAkito** [reply]
For clarification, my point of view.

### Module Master
```nu
export alias ll = ^ls -l
```

### Module Config
```nu
use /home/user/.config/nushell/master.nu *
```

### My Nu Shell Environment
```nu
/home/user/whatever> source /home/user/.config/nushell/config.nu
/home/user/whatever> ll # <- Why is this available in my environment, now?
```

**Reply by theAkito** [reply]
Okay, I think I get it. Sourcing reads the file verbatim, as if it were executed in my terminal, so my terminal essentially executes `use /home/user/.config/nushell/master.nu *`. Although, this is the same behaviour, as in Bash and I thought I read somewhere in the documentation, that the behaviour were different. 🤔

**Reply by kubouch** [reply]
Yes, you're right, `source` literally pastes the code from the script and executes it.

The difference vs. Bash is that Nushell is a "compiled" language: First it parses the source, then executes it. There is no "eval" functionality like in Bash and other dynamic languages. This means that you cannot source a dynamic path: All the sources need to be ready at parse time, before you start evaluating, similar to C's `#include` for example.

**Reply by theAkito** [reply]
Okay, that makes sense.
Thanks for the explanation, I understand now. 😄
